neps/nep-0522.md

```diff
@@ -0,0 +1,203 @@
+
+## Specification
+
+We describe `Transaction` and `DelegateAction`s collectively as messages in this specification.
+
+We propose to add the optional fields to the messages:
+
+```rust
+expires_at: Option<u64>,
+random_nonce: Option<u64>,
+```
+
+You first construct a trie of the following type where `|` represents a tagged sum type:
+
+```
+Expiry In Seconds | Expiry Block Height => Hash Transaction | Hash DelegateAction => ()
+```
+
+When the `Transaction` or `DelegateAction` is received it's Borsh representation is hashed. If `expires_at` is used the expiry is rounded to the nearest second. When `max_block_height`is used to describe expiry it's directly inserted.
+
+We then ensure that the message is unique using the following pseudocode:
+
+```rust
+fn ensure_unique(trie: &mut Trie, message: Transaction | DelegateAction) {
+    // Does is have expires_at or a valid max_block_height?
+    let expiry: MaxBlockHeight | ExpiresAt = message.expiry_type();
+    // If it has an expiry time ceil it to the next second
+    if expiry is ExpiresAt {
+        expiry = epiry.ceil_to_second()
+    }
+    // Insert into the trie
+    let already_exists = trie.insert(expiry, message.hash());
+    assert_eq!(exists, false);
+}
+```
+
+At the end of every block we need to cleanup the trie, making sure not to remove any hashes that might still be valid.
+
+```rust
+fn cleanup_trie(trie: &mut Trie, chunk_before_last: Chunk, last_chunk: Chunk) {
+    let blocktime_before_last = chunk_before_last.block_time.ceil_to_second();
+    let last_blocktime = chunk_before_last.block_time.ceil_to_second();
+
+    // Get all the seconds where things have expired since the last cleanup
+    // This doesn't include the last blocktime (.. not ..=)
+    let expired_times = blocktime_before_last..last_blocktime;
+
+    for t in expired_times {
+        // Remove all hashes that expired at this time
+        trie.remove_all(ExpiresAt(t));
+    }
+
+    // Remove all hashes that expired at the previous block
+    trie.remove_all(MaxBlockHeight(last_chunk.block_height))
+}
+```
+
+The nonce only needs to exist on the shard containing the sender account since that is the only place it can be sent from. On a shard split we need to ensure that all valid transaction hashes are sent to both of the new shards[^3].
+
+If a client attempts to send an identical transaction with an identical `random_nonce` we preserve the existing behavior for responding to an already sent transaction. Provided it still exists they will receive the previous transactions receipt and no action will be taken on chain. 
+
+There is no requirement that the `random_nonce` field is unique or random. A client could (inadvisably) decide to always provide a `random_nonce` with a value of 0 and it would work as expected until they tried to send two completely identical transactions. 
```

Comment:
We'll need to make sure this doesn't conflict with plans around transaction priority / gas wars.  On many other chains at least, there's an explicit feature to allow the sender to overwrite a pending transaction at a given key+nonce with a new one, either to attempt to proactively 'cancel' a transaction that isn't likely to be included in a block or to increase it's priority so it'll get included faster.

That first use case isn't as important here, because a pending transaction with a given `random_nonce` isn't preventing other transactions with different `random_nonce`s like it would for the incrementing `nonce`s.  But transaction priority features might require that we change this part of the spec.

Reply:
This doesn't prevent any of these mechanisms when not using a `random_nonce`.

Also if we want people to be able to cancel/re-submit outstanding transactions, wouldn't it be better to do it for individual transactions rather than all outstanding transactions using nonce contention? It lead to a nicer API and fewer re-submissions during peak load.